src/v/storage/mvlog/batch_collecting_stream_utils.h

```diff
@@ -0,0 +1,42 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "storage/mvlog/entry_stream.h"
+#include "storage/mvlog/reader_outcome.h"
+
+namespace storage::experimental::mvlog {
+
+class batch_collector;
+class entry_stream;
+
+enum class collect_stream_outcome {
+    // The batch collector is full.
+    buffer_full,
+
+    // There are no more entries in the stream.
+    end_of_stream,
+
+    // The batch collection is complete, e.g. because we have reached the
+    // desired offset.
+    stop,
+};
+std::ostream& operator<<(std::ostream&, collect_stream_outcome);
+
+// Parses and collects the record batches from the given entry stream.
+// Ignores other kinds of entries.
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream& entries, batch_collector& collector);
```

Comment:
should we pass raw pointers to avoid clang-tidy warnings on reference parameters to coroutines?

Reply:
Ah that reminds me this is actually partially the reason I had pointers in the first place. I opted to change them to abide by GSG, as called out by others, though maybe it's worth an exception for coroutines... 🤔

Reply:
@dotnwat the benefit would be that the need for explicit `&` will make you think twice about the lifetimes? Or something else?

Reply:
I guess it can also prevent cases where a temporary would be otherwise destroyed?

i.e. `co_await collect_batches_from_stream(entry_stream{input}, c)`? does entry_stream here gets destroyed on suspension?

Reply:
Yeah the clang-tidy warning is a good callout. Allowing passing a temporary as a reference is a crash waiting to happen.

Reply:
Anyone can word it out nicely for https://github.com/redpanda-data/redpanda/blob/dev/src/v/coding-style.md?

Reply:
right, pointer is cleaner for coroutine. i mean we could also disable the clang-tidy warning.